### PR TITLE
Use mutating webhook to inject sidecar

### PR DESCRIPTION
**Pull request by lujiajing1126**
## Which problem is this PR solving?
- #171 
- Also #1232 

## Short description of the changes

Based on [the idiom of istio injection policy](https://istio.io/latest/docs/setup/additional-setup/sidecar-injection/#controlling-the-injection-policy), 

| Resource  | Label | Enabled value | Disabled value |
| ------------- | ------------- | ------------- | ------------- |
| Namespace | jaeger-injection | enabled | disabled |
| Pod | sidecar.jaertracing.io/inject | "true" | "false" |

This idiom could safely prevent from [bootstrapping problem](https://book.kubebuilder.io/cronjob-tutorial/running-webhook.html#the-bootstrapping-problem) supported from Kubernetes v1.15+

- `CertManager` should be installed to inject CA certification as is sugggested by [`kubebuilder-book`](https://book.kubebuilder.io/cronjob-tutorial/cert-manager.html)

## Tasks

- [x] Add webhook and minifests
- [x] E2E tests: TBD

## To be discussed

1. Shall we totally remove namespace controller and deployment controller or just introduce this feature as an experimental thing in the next few releases keeping two kinds of injection approaches in parallel?
2. Annotation based injection is not currently supported by this webhook.

### CODE REVIEW

**Comment by lujiajing1126** [comment]
Though more test cases are needed and issue #1692 has to be resolved first, generally ready to be reviewed.

Request review @jpkrohling @rubenvp8510

**Comment by jpkrohling** [comment]
@rubenvp8510 and @pavolloffay would be the appropriate reviewers for this.

**Comment by rubenvp8510** [comment]
I would like to get rid of namespace and deployment controllers at some point, but the only concern I have with this is it seems to be a breaking change from what the users are doing now to inject the sidecar.

Any reason to move it to Pod?  Is because it provides more flexibility?

File: .github/workflows/e2e.yaml [hunk 1]
```diff
@@ -199,6 +199,33 @@ jobs:
         env:
           KUBE_VERSION: ${{ matrix.kube-version }}
         run: make run-e2e-tests-outside-cluster
+  sidecard-e2e-tests:
+    name: Smoke E2E tests
```
Comment (rubenvp8510) [inline .github/workflows/e2e.yaml:203]: Is this name correct? Should we call this Smoke sidecar e2e tests?

File: config/webhook/component_inject_patch.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    objectSelector:
+      matchExpressions:
+        - key: app
+          operator: In
+          values:
+            - "jaeger"
+        - key: app.kubernetes.io/component
+          operator: NotIn
+          values:
+            - "query"
```
Comment (rubenvp8510) [inline config/webhook/component_inject_patch.yaml:16]: new line

File: config/webhook/kustomization.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+  - manifests.yaml
+  - service.yaml
+
+configurations:
+  - kustomizeconfig.yaml
+
+patchesStrategicMerge:
+  - namespace_inject_patch.yaml
+  - pod_inject_patch.yaml
+  - component_inject_patch.yaml
```
Comment (rubenvp8510) [inline config/webhook/kustomization.yaml:11]: add new line at the end

File: config/webhook/namespace_inject_patch.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+          values:
+            - "enabled"
+    objectSelector: # also skip pods that have been already injected (prevent from being injected twice by multiple operators)
+      matchExpressions:
+        - key: sidecar.jaegertracing.io/inject
+          operator: NotIn
+          values:
+            - "false"
+        - key: sidecar.jaegertracing.io/injected # set by deployment controller
+          operator: DoesNotExist
```
Comment (rubenvp8510) [inline config/webhook/namespace_inject_patch.yaml:20]: new line

**Comment by lujiajing1126** [comment]
> I would like to get rid of namespace and deployment controllers at some point, but the only concern I have with this is it seems to be a breaking change from what the users are doing now to inject the sidecar.

I cannot agree more about this, "namespace and deployment controllers" have to be removed.

After migrating to the pod mutating webhook, the current injected sidecar will not be removed. Mutating webhook just provides a new way to do so which is orthogonal to the previous one. As I understand, it is the backwards compatibility guarantee.

> Any reason to move it to Pod? Is because it provides more flexibility?

I suppose there are multiple reasons, the most important of which is stable. We have met several issues in the production cluster with deployment/namespace reconciliation,

1. Issue #1696 is caused by deployment automatic reconciliation. This issue made surge in our production cluster twice, leading to thousands of pods creation within a small interval.
2. Issue #1318 is also caused by deployment reconciliation which had been fixed later.

Another issue is the upgrade: performing upgrade for the operator has to be extremely careful since it will update all deployments to inject new sidecar. It would be very scary in a relatively large cluster.

**Comment by rubenvp8510** [comment]
@lujiajing1126  Sorry for been late on this, but this PR: https://github.com/jaegertracing/jaeger-operator/pull/1718 has been merged already, I think you needed in order to test this, right?

**Comment by frzifus** [comment]
hi @lujiajing1126, would you like to continue working on this topic? alternatively, i would like to continue your work.

**Comment by lujiajing1126** [comment]
> hi @lujiajing1126, would you like to continue working on this topic? alternatively, i would like to continue your work.

I am too busy to proceed on this PR. Sure please go ahead!

File: config/default/webhookcainjection_patch.yaml [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# This patch add annotation to admission webhook config and
+# the variables $(CERTIFICATE_NAMESPACE) and $(CERTIFICATE_NAME) will be substituted by kustomize.
+apiVersion: admissionregistration.k8s.io/v1
+kind: MutatingWebhookConfiguration
+metadata:
+  name: mutating-webhook-configuration
+  annotations:
+    cert-manager.io/inject-ca-from: $(CERTIFICATE_NAMESPACE)/$(CERTIFICATE_NAME)
+#---
```
Comment (iblancasa) [inline config/default/webhookcainjection_patch.yaml:9]: Is it OK to keep this commented code?

File: pkg/inject/sidecar.go [hunk 6]
```diff
@@ -19,10 +19,17 @@ import (
...
+	// Only "enabled" is the valid value for namespace-scoped injection
+	NamespaceLabel = "jaeger-injection"
+	// Annotation is the annotation name to look for when deciding whether to inject
 	Annotation = "sidecar.jaegertracing.io/inject"
-	// Label is the label name the operator put on injected deployments.
+	// Label is the label name the operator put on injected deployments or pods.
 	Label = "sidecar.jaegertracing.io/injected"
+	// PodLabel is the label name on the pod level to look for when deciding whether to inject
+	// This is required and recommended approach to make use of mutating-webhook
+	// TODO: support sidecar injection via annotation-level declaration like istio
```
Comment (iblancasa) [inline pkg/inject/sidecar.go:31]: Create an issue to address it in the future and add here the issue ID.

**Review by iblancasa — CHANGES_REQUESTED** [review]

**Comment by rubenvp8510** [comment]
I'll close this PR , as was superseded by https://github.com/jaegertracing/jaeger-operator/pull/1828

Thanks for this PR which was used as the base for the webhook work!

**Pull request closed without merging**